### PR TITLE
Added NuGet.Services.Status.Table (table storage classes)

### PR DESCRIPTION
**Pull request by scottbommarito**
Also removed interfaces from `Event` and `Message` because they had no value.

### CODE REVIEW

File: src/NuGet.Services.Status/Event.cs [hunk 1]
```diff
@@ -7,42 +7,49 @@
...
+
+        /// <summary>
+        /// The <see cref="IReadOnlyComponent.Status"/> of the <see cref="IReadOnlyComponent"/> affected.
+        /// </summary>
         public ComponentStatus AffectedComponentStatus { get; }
+
+        /// <summary>
+        /// When the event began.
+        /// </summary>
         public DateTime StartTime { get; }
```
Comment (loic-sharma) [inline src/NuGet.Services.Status/Event.cs]: >DateTime [](start = 15, length = 8)

If possible, you should prefer DateTimeOffset

File: src/NuGet.Services.Status.Table/EventEntity.cs [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+            : this(incidentEntity.AffectedComponentPath, incidentEntity.AffectedComponentStatus, incidentEntity.CreationTime)
+        {
+            incidentEntity.EventRowKey = RowKey;
+        }
+
+        public string AffectedComponentPath { get; set; }
+        public int AffectedComponentStatus { get; set; }
+        public DateTime StartTime { get; set; }
+        public DateTime? EndTime { get; set; }
+        public bool IsActive
```
Comment (loic-sharma) [inline src/NuGet.Services.Status.Table/EventEntity.cs]: >IsActive [](start = 20, length = 8)

Can this be just `IsActive => EndTime == null;`? What's the setter for?

File: src/NuGet.Services.Status/Event.cs (hunk 1 above)
Comment (scottbommarito) [reply src/NuGet.Services.Status/Event.cs]: Azure Table storage only supports `DateTime` unfortunately

File: src/NuGet.Services.Status.Table/EventEntity.cs (hunk 2 above)
Comment (scottbommarito) [reply src/NuGet.Services.Status.Table/EventEntity.cs]: Basically, `TableEntity` (base class of this) has a serializer that only serializes a property if it has both a getter and setter. I'm doing this to trick it into serializing this property.

File: src/NuGet.Services.Status.Table/CursorEntity.cs [hunk 3]
```diff
@@ -0,0 +1,23 @@
+﻿using Microsoft.WindowsAzure.Storage.Table;
+using System;
```
Comment (chenriksson) [inline src/NuGet.Services.Status.Table/CursorEntity.cs]: nit: sort, and add copyright

File: src/NuGet.Services.Status.Table/CursorEntity.cs [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+namespace NuGet.Services.Status.Table
+{
+    public class CursorEntity : TableEntity
+    {
+        public const string DefaultPartitionKey = "cursors";
+        public const string DefaultRowKey = "1";
+
+        public CursorEntity()
+        {
+        }
```
Comment (chenriksson) [inline src/NuGet.Services.Status.Table/CursorEntity.cs]: used?

File: src/NuGet.Services.Status.Table/EventEntity.cs [hunk 5]
```diff
@@ -0,0 +1,71 @@
+﻿using Microsoft.WindowsAzure.Storage.Table;
+using System;
+using System.Collections.Generic;
```
Comment (chenriksson) [inline src/NuGet.Services.Status.Table/EventEntity.cs]: nit: sort and copyright

File: src/NuGet.Services.Status.Table/EventEntity.cs [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    /// <summary>
+    /// Class used to serialize an <see cref="Event"/> in a table.
+    /// </summary>
+    public class EventEntity : TableEntity
+    {
+        public const string DefaultPartitionKey = "events";
+
+        public EventEntity()
+        {
+        }
```
Comment (chenriksson) [inline src/NuGet.Services.Status.Table/EventEntity.cs]: used?

File: src/NuGet.Services.Status.Table/EventEntity.cs [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+        public EventEntity(IncidentEntity incidentEntity)
+            : this(incidentEntity.AffectedComponentPath, incidentEntity.AffectedComponentStatus, incidentEntity.CreationTime)
+        {
+            incidentEntity.EventRowKey = RowKey;
+        }
+
+        public string AffectedComponentPath { get; set; }
+        public int AffectedComponentStatus { get; set; }
```
Comment (chenriksson) [inline src/NuGet.Services.Status.Table/EventEntity.cs]: Why cast to int? Is `ComponentStatus` not exposed to callers?

File: src/NuGet.Services.Status.Table/EventEntity.cs [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+        public DateTime StartTime { get; set; }
+        public DateTime? EndTime { get; set; }
+        public bool IsActive
+        {
+            get { return EndTime == null; }
+            set { }
+        }
```
Comment (chenriksson) [inline src/NuGet.Services.Status.Table/EventEntity.cs]: nit: do we typically sort properties before ctors?

File: src/NuGet.Services.Status.Table/EventEntity.cs [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+
+        public string AffectedComponentPath { get; set; }
+        public int AffectedComponentStatus { get; set; }
+        public DateTime StartTime { get; set; }
+        public DateTime? EndTime { get; set; }
+        public bool IsActive
+        {
+            get { return EndTime == null; }
+            set { }
```
Comment (chenriksson) [inline src/NuGet.Services.Status.Table/EventEntity.cs]: Can you remove the empty setter? If it's there for serialization, can't you add `JsonIgnore` or similar attribute?

File: src/NuGet.Services.Status.Table/IncidentEntity.cs [hunk 10]
```diff
@@ -0,0 +1,49 @@
+﻿using Microsoft.WindowsAzure.Storage.Table;
+using System;
```
Comment (chenriksson) [inline src/NuGet.Services.Status.Table/IncidentEntity.cs]: nit: sort and copyright

File: src/NuGet.Services.Status.Table/IncidentEntity.cs [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+    /// <summary>
+    /// Class used to serialize an incident in a table.
+    /// </summary>
+    public class IncidentEntity : TableEntity
+    {
+        public const string DefaultPartitionKey = "incidents";
+
+        public IncidentEntity()
+        {
+        }
```
Comment (chenriksson) [inline src/NuGet.Services.Status.Table/IncidentEntity.cs]: used?

File: src/NuGet.Services.Status.Table/IncidentEntity.cs [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+
+        public IncidentEntity(string id, string affectedComponentPath, ComponentStatus affectedComponentStatus, DateTime creationTime, DateTime? mitigationTime)
```
Comment (chenriksson) [inline src/NuGet.Services.Status.Table/IncidentEntity.cs]: nit: wrap long lines?

File: src/NuGet.Services.Status.Table/IncidentEntity.cs [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+            MitigationTime = mitigationTime;
+        }
+
+        public string EventRowKey { get; set; }
+        public bool IsLinkedToEvent
+        {
+            get { return !string.IsNullOrEmpty(EventRowKey); }
+            set { }
+        }
+        public string IncidentApiId { get; set; }
```
Comment (chenriksson) [inline src/NuGet.Services.Status.Table/IncidentEntity.cs]: Why not just `IncidentId`? What does added `Api` signify?

File: src/NuGet.Services.Status.Table/IncidentEntity.cs [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+            AffectedComponentStatus = (int)affectedComponentStatus;
+            CreationTime = creationTime;
+            MitigationTime = mitigationTime;
+        }
+
+        public string EventRowKey { get; set; }
+        public bool IsLinkedToEvent
+        {
+            get { return !string.IsNullOrEmpty(EventRowKey); }
+            set { }
```
Comment (chenriksson) [inline src/NuGet.Services.Status.Table/IncidentEntity.cs]: Remove empty setter?

File: src/NuGet.Services.Status.Table/IncidentEntity.cs [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+        }
+        public string IncidentApiId { get; set; }
+        public string AffectedComponentPath { get; set; }
+        public int AffectedComponentStatus { get; set; }
+        public DateTime CreationTime { get; set; }
+        public DateTime? MitigationTime { get; set; }
+        public bool IsActive
+        {
+            get { return MitigationTime == null; }
+            set { }
```
Comment (chenriksson) [inline src/NuGet.Services.Status.Table/IncidentEntity.cs]: Remove empty setter?

File: src/NuGet.Services.Status.Table/MessageEntity.cs [hunk 16]
```diff
@@ -0,0 +1,39 @@
+﻿using Microsoft.WindowsAzure.Storage.Table;
+using System;
```
Comment (chenriksson) [inline src/NuGet.Services.Status.Table/MessageEntity.cs]: nit: sort and copyright

File: src/NuGet.Services.Status.Table/CursorEntity.cs (hunk 4 above)
Comment (scottbommarito) [reply src/NuGet.Services.Status.Table/CursorEntity.cs]: required by `TableEntity`

File: src/NuGet.Services.Status.Table/EventEntity.cs (hunk 7 above)
Comment (scottbommarito) [reply src/NuGet.Services.Status.Table/EventEntity.cs]: Table storage ignores enum objects in serialization. I use int here to make the conversion easy.

File: src/NuGet.Services.Status.Table/IncidentEntity.cs (hunk 13 above)
Comment (scottbommarito) [reply src/NuGet.Services.Status.Table/IncidentEntity.cs]: My thoughts were to make it specific that this was the ID returned from the API. Table storage has its own "id" for each entity, e.g. partition key and row key, and I wanted to make it clear that this was the ID from the API.

File: src/NuGet.Services.Status.Table/EventEntity.cs (hunk 9 above)
Comment (scottbommarito) [reply src/NuGet.Services.Status.Table/EventEntity.cs]: https://github.com/NuGet/ServerCommon/pull/176#discussion_r204564192

Also, `TableEntity` serialization doesn't use `JSON.NET`, so unfortunately attributes don't have any effect.

**Comment by scottbommarito** [comment]
@chenriksson I've added some comments that should help with your confusion.

File: src/NuGet.Services.Status.Table/EventEntity.cs (hunk 6 above)
Comment (scottbommarito) [reply src/NuGet.Services.Status.Table/EventEntity.cs]: required by `TableEntity`

File: src/NuGet.Services.Status.Table/IncidentEntity.cs (hunk 11 above)
Comment (scottbommarito) [reply src/NuGet.Services.Status.Table/IncidentEntity.cs]: required by `TableEntity`

File: src/NuGet.Services.Status.Table/CursorEntity.cs [hunk 17]
```diff
@@ -0,0 +1,26 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.WindowsAzure.Storage.Table;
+using System;
```
Comment (chenriksson) [inline src/NuGet.Services.Status.Table/CursorEntity.cs]: nit: sort system first; same below

**Comment by scottbommarito** [comment]
Merging as soon as I've tested this library on DEV with the status aggregator job.

**Pull request merged**